src/main/java/net/minestom/server/entity/ai/EntityAIGroup.java

```diff
@@ -0,0 +1,143 @@
+package net.minestom.server.entity.ai;
+
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.UnaryOperator;
+
+/**
+ * Represents a group of entity's AI.
+ * It may contains {@link GoalSelector goal selectors} and {@link TargetSelector target selectors}.
+ * All AI groups of a single entity are independent of each other.
+ */
+public class EntityAIGroup {
+
+    private GoalSelector currentGoalSelector;
+    private final List<GoalSelector> goalSelectors = new GoalSelectorsArrayList();
+    private final List<TargetSelector> targetSelectors = new ArrayList<>();
+
+    /**
+     * Gets the goal selectors of this group.
+     *
+     * @return a modifiable list containing this group goal selectors
+     */
+    @NotNull
+    public List<GoalSelector> getGoalSelectors() {
+        return this.goalSelectors;
+    }
+
+    /**
+     * Gets the target selectors of this group.
+     *
+     * @return a modifiable list containing this group target selectors
+     */
+    @NotNull
+    public List<TargetSelector> getTargetSelectors() {
+        return this.targetSelectors;
+    }
+
+    /**
+     * Gets the current goal selector of this group.
+     *
+     * @return the current goal selector of this group, null if not any
+     */
+    @Nullable
+    public GoalSelector getCurrentGoalSelector() {
+        return this.currentGoalSelector;
+    }
+
+    /**
+     * Changes the current goal selector of this group.
+     * <p>
+     * Mostly unsafe since the current goal selector should normally
+     * be chosen during the group tick method.
+     *
+     * @param goalSelector the new goal selector of this group, null to disable it
+     */
+    public void setCurrentGoalSelector(@Nullable GoalSelector goalSelector) {
+        Check.argCondition(
+                goalSelector != null && goalSelector.getAIGroup() != this,
+                "Tried to set goal selector attached to another AI group!"
+        );
+        this.currentGoalSelector = goalSelector;
+    }
+
+    public void tick(long time) {
+        GoalSelector currentGoalSelector = getCurrentGoalSelector();
```

Comment:
When possible, try to avoid reusing the same name than a field.
It's by no mean a big issue but it did confuse me a bit a few lines down.